:book: clarify documentation, add more examples (#134)
itpp-labs/sync-addons: openapi/doc/index.rst

--------------

* `Activate Developer Mode <https://odoo-development.readthedocs.io/en/latest/odoo/usage/debug-mode.html>`__
* Open menu ``[[ Settings ]] >> Users & Companies >> Users``
* Select a user that will be used for iteracting over API
* In **Allowed Integration** select some integrations
* Copy **OpenAPI Token** to use it in any system that support OpenAPI

> If necessary, you can reset the token pressing ``Reset OpenAPI Token``

Reviewer: ```suggestion
If necessary, you can reset the token by pressing ``[Reset OpenAPI Token]`` button
```

Usage
=====

Swagger Editor
--------------
As the simplest example, you can try API in Swagger Editor. It allows to review and check API

* Open http://editor.swagger.io/

> * Click menu ``File >> Import URL``

Reviewer: :+1:

* Set **Specification link**
* RESULT: Specification is parsed succefully and you can see API presentation
* Click ``[Authorize]`` button

  * **Username** -- set database name

>   * **Password** -- set **OpenAPI Token** (described `here <#authentication>`__)

Reviewer: ```suggestion
  * **Password** -- set **OpenAPI Token** (how to get one is described in `authentication <#authentication>`__ above)
```

Note:
  The Swagger Editor sends requests directly from browser which leads to CORS error and work with it is not available in `odoo.sh`.
  The easiest solution is to simply copy-past the curl command from Swagger Editor and run it from the terminal.

>   If you're using Nginx you can add a setting like this::

Reviewer: ```suggestion
  Alternatively, you can grant CORS headers in your web server. Below is example for Nginx::
```

    location /api/v1 {
       if ($request_method = 'OPTIONS') {
          add_header 'Access-Control-Allow-Origin' '*';
          add_header 'Access-Control-Allow-Credentials' 'true';
          add_header 'Access-Control-Max-Age' 1728000;
          add_header 'Content-Type' 'text/plain charset=UTF-8';
          add_header 'Content-Length' 0;
          return 204;
       }
       if ($request_method = 'POST') {
          add_header 'Access-Control-Allow-Origin' '*';
          add_header 'Access-Control-Allow-Credentials' 'true';
       }
       if ($request_method = 'GET') {
          add_header 'Access-Control-Allow-Origin' '*';
          add_header 'Access-Control-Allow-Credentials' 'true';
       }
    }

How to call methods with arguments via API
------------------------------------------

Here is an example of calling a search method with domain.

> This is how it is usually done from the odoo model:

Reviewer: ```suggestion
This is how it is usually done from python code:
```

.. code-block:: python

  partner_ids = self.env['res.partner'].search([("is_company", "=", "True")])

> Using API:

Reviewer: ```suggestion
On using API it would be as following:
```

  * **Username** -- set database name
  * **Password** -- set **OpenAPI Token**

How to call methods with arguments via API
------------------------------------------

Here is an example of calling a search method with domain.
This is how it is usually done from the inside:

.. code-block:: python

  partner_ids = self.env['res.partner'].search([("is_company", "=", "True")])

Using API:

.. code-block:: bash

  curl -X PATCH "http://example.com/api/v1/demo/res.partner/call/search" -H "accept: application/json" \
  -H "authorization: Basic BASE64_ENCODED_EXPRESSION" -H "Content-Type: application/json" \
  -d '{ "args": [[["is_company", "=", "True" ]]]}'

Get OpenAPI Token for user
--------------------------

* Open menu ``[[ Settings ]] >> Users & Companies >> Users``
* Select a user that will be used for iteracting over API (Added to `Allowed users`)

> * Open ``Allowed Integrations``, from here you can copy the token

Reviewer: Это уже есть выше

```

* Open menu ``[[ Settings ]] >> Users >> Users``
* Select a user that will be used for iteracting over API
* In **Allowed Integration** select some integrations
* Copy **OpenAPI Token** to use it in any system that support OpenAPI
```

Author: Тут не понятно откуда его нужно брать. Я сам с этим столкнулся, когда в первый раз использовал.

Author: Хотя да, почему то мимо этого заголовка постоянно проходил
